lib/trino-filesystem/src/main/java/io/trino/filesystem/fileio/BulkOperationsFileIo.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.fileio;
+
+import io.trino.filesystem.TrinoFileSystem;
+import org.apache.iceberg.io.BulkDeletionFailureException;
+import org.apache.iceberg.io.SupportsBulkOperations;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class BulkOperationsFileIo
+        extends ForwardingFileIo
+        implements SupportsBulkOperations
+{
+    private static final int DELETE_BATCH_SIZE = 1000;
+
+    public BulkOperationsFileIo(TrinoFileSystem fileSystem)
+    {
+        super(fileSystem);
+    }
+
+    @Override
+    public void deleteFiles(Iterable<String> pathsToDelete)
+            throws BulkDeletionFailureException
+    {
+        List<String> filesToDelete = new ArrayList<>();
```

Comment:
Please use `ImmutableList.Builder` instead of `ArrayList`.
Also I think you can replace this logic here with something like:

```
Iterable<List<String>> partitions = Iterables.partition(pathsToDelete, DELETE_BATCH_SIZE);
for (List<String> currentBatch : partitions) {
    deleteBatch(currentBatch);
}
```

Reply:
`ArrayList` better over `ImmutableList.Builder` because of re-allocations (https://github.com/trinodb/trino/pull/14434/files#diff-e1cb17efec6787989f9df9ee40c4f2809ff3fe946cd2ec721ff8932b131997b8R1183)

👍 for `Iterables.partition`